src/TensorFlowNET.Core/Checkpoint/functional_saver.cs

```diff
@@ -0,0 +1,559 @@
+﻿using System;
+using System.Buffers.Text;
+using System.Collections.Generic;
+using System.Text;
+using Tensorflow.Train;
+using static Tensorflow.ApiDef.Types;
+using static Tensorflow.CostGraphDef.Types;
+using static Tensorflow.OptimizerOptions.Types;
+using static Tensorflow.Binding;
+using System.Text.RegularExpressions;
+using System.Linq;
+using Tensorflow.Operations;
+using Tensorflow.Training;
+using Tensorflow.Graphs;
+using System.Xml.Linq;
+using System.Diagnostics;
+
+namespace Tensorflow.Checkpoint
+{
+    /// <summary>
+    /// `FunctionHolder` is a series of containers to help dynamically call some dotnet functions.
+    /// Note that this API does not gurantee performance. Besides, it is not supposed to be exposed to users.
+    /// </summary>
+    public interface IFunctionHolder
+    {
+        int ArgCount { get; }
+        object DynamicInvoke(params object[] args);
+    }
+    internal record class FunctionHolder<TR>(Func<TR> Func): IFunctionHolder
+    {
+        public int ArgCount => 0;
+        public object DynamicInvoke(params object[] args)
+        {
+            return Func.DynamicInvoke(args);
+        }
+        public TR Invoke()
+        {
+            return Func.Invoke();
+        }
+    }
+    internal record class FunctionHolder<TA1, TR>(Func<TA1, TR> Func) : IFunctionHolder
+    {
+        public int ArgCount => 1;
+        public object DynamicInvoke(params object[] args)
+        {
+            return Func.DynamicInvoke(args);
+        }
+    }
+    internal record class FunctionHolder<TA1, TA2, TR>(Func<TA1, TA2, TR> Func) : IFunctionHolder
+    {
+        public int ArgCount => 2;
+        public object DynamicInvoke(params object[] args)
+        {
+            return Func.DynamicInvoke(args);
+        }
+    }
+    internal record class FunctionHolder<TA1, TA2, TA3, TR>(Func<TA1, TA2, TA3, TR> Func) : IFunctionHolder
+    {
+        public int ArgCount => 3;
+        public object DynamicInvoke(params object[] args)
+        {
+            return Func.DynamicInvoke(args);
+        }
+    }
+    public class Maybe<TA, TB>
```

Comment:
@Oceania2018 Please help to review this class. It seems to have an ugly design.😂 What I want is a class that allows to be either of two types. The two type often do not inherit from common interface. Using `object` is a solution but it does not have a strong constraint of type.

Reply:
Can we just remove `FunctionHolder`? use `lambda` instead?

Reply:
The case that the number of input args is uncertain is one of the reasons of designing `Maybe`. The other is that sometimes a return value of a function may be one type or another. If I separate the function to two, the place which call the function has to deal with two return types. The more the nest of calls, the more extra work it needs. 

Thank you for your advice and I've removed `Functionholder`.